adafruit_ads1x15/analog_in.py

```diff
@@ -55,12 +55,49 @@ def value(self) -> int:
         Even if the underlying analog to digital converter (ADC) is
         lower resolution, the value is 16-bit.
         """
-        return self._ads.read(
-            self._pin_setting, is_differential=self.is_differential
-        ) << (16 - self._ads.bits)
+        return self._ads.read(self._pin_setting, is_differential=self.is_differential)
 
     @property
     def voltage(self) -> float:
         """Returns the voltage from the ADC pin as a floating point value."""
-        volts = self.value * _ADS1X15_PGA_RANGE[self._ads.gain] / 32767
+        volts = self.convert_to_voltage(self.value)
+        return volts
+
+    def convert_to_value(self, volts: float) -> int:
+        """Calculates 12-bit integer for threshold registers from voltage level input"""
+
+        # Convert 2's complement of signed int if number is negative
+        if volts > 0:
+            value = round(
+                (volts / _ADS1X15_PGA_RANGE[self._ads.gain])
+                * ((1 << (self._ads.bits - 1)) - 1)
+            )
+        else:
+            value = round(
+                (volts / _ADS1X15_PGA_RANGE[self._ads.gain])
+                * (1 << (self._ads.bits - 1))
+            )
+            value += 1 << self._ads.bits
+
+        # Need to bit shift if value is only 12-bits
+        value <<= 16 - self._ads.bits
+        return value
+
+    def convert_to_voltage(self, value_int: int) -> float:
+        """Calculates voltage from 16-bit ADC reading"""
+
+        if value_int & 0x8000:
+            # Need to convert negative number through 2's complement
+            value_int -= 0x10000
+
+        # Need to bit shift if value is only 12-bits
+        value_int >>= 16 - self._ads.bits
```

Comment:
Again. Don't worry about two's complement or the numbers of bits. This should be 16 bit unsigned or "17 bit" signed to voltage.

Reply:
I think we are on different pages as far as how to represent value. From "Signed number representations" on Wikipedia, 8-bit signed goes from -127 to 127. 8-bit unsigned goes from 0 to 255. I think value should match the registers in the datasheet, which is 16 bit two's complement or signed.

Can you explain 17-bit to me?

Voltage is a float since we want it to have decimal places.

Reply:
value for reading from an adc has been defined as 16-bit unsigned since we added the AnalogIn API. If you want this resolution signed, then you need to add another bit for the sign. That would make it 17 total bits. I realize that the comparator will only do 16-bit signed. This conversion from voltage to value should handle the full 16-bit unsigned range still though.